docker/archive/multi-reader.go

```diff
@@ -0,0 +1,77 @@
+package archive
+
+import (
+	"context"
+
+	"github.com/containers/image/v5/docker/tarfile"
+	"github.com/containers/image/v5/types"
+)
+
+// MultiImageArchiveReader allows for reading a docker-archive with multiple
+// images.  It can be used in subsequent copy.Image operations when set as the
+// copy.Options.ImageSource.  Note that you need to call Next() before each
+// copy.Image operation.
```

Comment:
I don’t think this needs to be, or should be, that stateful.

Instead of embedding an `archiveImageSource`, and trying to “reset” it, this could basically wrap `docker/tarfile.Source.{tarPath,removeTarPathOnClose}` (+`openTarComponent`), and provide a way to list the manifests and just get `docker/archive.archiveReference` values that are “bound” to that `tarPath`.

That
- would allow reading from the archive in parallel
- would be clearly correct in not sharing state across images in the archive
- would force use to deal with the `ImageReference` values, not just using the right tags, but dealing with untagged images and introducing names for them (which IIRC can be created) — and that would also allow referencing any _single_ image using the string `docker-archive:` syntax by tag / untagged.

Reply:
I like your proposal. Note that I very likely will run out of time for extending the `docker-archive` syntax but we would definitely get the ground work done.